File: ggcommons/config/manager/config_manager.py
import abc
import logging
import os
import sys
import time
from abc import abstractmethod

from ggcommons.config.heartbeat_config import HeartbeatConfiguration
from ggcommons.config.metric_config import MetricConfiguration
from ggcommons.config.tag_config import TagConfiguration
from ggcommons.config.logging_config import LoggingConfiguration

# import unittest

logger = logging.getLogger("ConfigManager")


class ConfigManager(metaclass=abc.ABCMeta):
    def __init__(self, component_name: str):
        self._tag_config = None
        self._heartbeat_config = None
        self._metric_config = None
        self._component_config = None
        self._global_config = {}
        self._instances = {}
        self._change_listeners = []
        self._thing_name = (
            "NOT_GREENGRASS"
            if "AWS_IOT_THING_NAME" not in os.environ
            else os.environ["AWS_IOT_THING_NAME"]
        )
        self._component_name = component_name

    def init(self):
        config = self._load_configuration()
        if config is None:
            config = {"component": {}}
        self._apply_config(config)

    def _apply_config(self, config: dict):
        logging_json = None if "logging" not in config else config["logging"]
        self._logging_config = LoggingConfiguration(logging_json)
        logging.basicConfig(
            format=self._logging_config.get_format(),
            level=self._logging_config.get_level(),
        )
        logging.Formatter.converter = time.gmtime
        logging.StreamHandler(sys.stdout)

        tag_json = None if "tags" not in config else config["tags"]
        self._tag_config = TagConfiguration(tag_json)

        heartbeat_json = None if "heartbeat" not in config else config["heartbeat"]
        self._heartbeat_config = HeartbeatConfiguration(heartbeat_json)

        metric_json = None if "metricEmission" not in config else config["metricEmission"]
        self._metric_config = MetricConfiguration(metric_json)

        component_json = (
            {"global": {}, "instances": []}
            if "component" not in config
            else config["component"]
        )
        self._component_config = component_json
        self._global_config = (
            {}
            if "global" not in self._component_config
            else self._component_config["global"]
        )
        self._gen_instances_map()

    def _gen_instances_map(self):
        if "instances" in self._component_config:
            for instance in self._component_config["instances"]:
                self._instances[instance["id"]] = instance
                logger.debug(f"loaded config for {self._instances[instance['id']]}")

    def configuration_changed(self, new_config: dict) -> bool:
        logger.debug(f"configuration_changed: Applying new config: {new_config}")
        self._apply_config(new_config)

        logger.info(f"configuration_changed: Notifying change listeners")
        for listener in self._change_listeners:
            listener.on_configuration_change(new_config)
        return True

    def resolve_template(self, template: str) -> str:
        ret_val = template
        if "{ThingName}" in template:
            ret_val = ret_val.replace("{ThingName}", self._thing_name)
        if "{ComponentName}" in template:
            ret_val = ret_val.replace("{ComponentName}", self._component_name)
        tag_dict = (
            {} if self._tag_config is None else self._tag_config.to_dict()
        )
        for k in tag_dict.keys():
            key_template = "{" + k + "}"
            if key_template in template:
                ret_val = ret_val.replace(key_template, tag_dict[k])
        return ret_val

    @abstractmethod
    def _load_configuration(self) -> dict:
        pass

    def get_global_config(self) -> dict:
        return self._global_config

    def get_instance_ids(self) -> list:
        return [*self._instances]

    def get_instance_config(self, inst_id) -> dict:
        return self._instances[inst_id]

    def get_tag_config(self) -> TagConfiguration:
        return self._tag_config

    def get_heartbeat_config(self) -> HeartbeatConfiguration:
        return self._heartbeat_config

    def get_metric_config(self) -> MetricConfiguration:
        return self._metric_config

    def get_logging_config(self) -> LoggingConfiguration:
        return self._logging_config

    def get_thing_name(self) -> str:
        return self._thing_name

    def get_component_name(self) -> str:
        return self._component_name

    def add_config_change_listener(self, listener):
        self._change_listeners.append(listener)

    @abstractmethod
    def get_config_source(self) -> str:
        pass

File: src/main/java/com/aws/proserve/ggcommons/config/manager/ConfigManager.java
package com.aws.proserve.ggcommons.config.manager;

import com.github.cliftonlabs.json_simple.JsonArray;
import com.github.cliftonlabs.json_simple.JsonObject;
import com.aws.proserve.ggcommons.config.HeartbeatConfiguration;
import com.aws.proserve.ggcommons.config.LoggingConfiguration;
import com.aws.proserve.ggcommons.config.TagConfiguration;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

import java.util.ArrayList;
import java.util.Collection;
import java.util.HashMap;


public abstract class ConfigManager
{
    protected static final Logger LOGGER = LogManager.getLogger(ConfigManager.class);

    protected final String componentName;
    protected final String thingName;
    protected final ArrayList<ConfigurationChangeListener> configChangeListeners = new ArrayList<>();
    protected TagConfiguration tagConfig;
    protected HeartbeatConfiguration heartbeatConfig;
    protected JsonObject componentConfig;
    protected JsonObject globalConfig;
    protected LoggingConfiguration loggingConfig;
    protected final HashMap<String, JsonObject> instanceConfigs = new HashMap<>();


    ConfigManager(String componentName)
    {
        this.componentName = componentName;
        thingName = System.getenv("AWS_IOT_THING_NAME") != null ? System.getenv("AWS_IOT_THING_NAME") : "NOT_GREENGRASS";
    }

    protected void init()
    {
        JsonObject config = loadConfiguration();
        if (config != null)
        {
            applyConfig(config);
        }
    }

    private void applyConfig(JsonObject config)
    {
        loggingConfig = config.containsKey("logging") ? new LoggingConfiguration((JsonObject) config.get("logging")) : null;
        tagConfig = config.containsKey("tags") ? new TagConfiguration((JsonObject) config.get("tags")) : null;
        heartbeatConfig = config.containsKey("heartbeat") ? new HeartbeatConfiguration((JsonObject) config.get("heartbeat")) : null;

        componentConfig = (JsonObject) config.get("component");
        globalConfig = componentConfig.containsKey("global") ? (JsonObject) componentConfig.get("global") : null;
        genInstancesMap();
    }

    private void genInstancesMap()
    {
        JsonArray instances = componentConfig.containsKey("instances") ? (JsonArray) componentConfig.get("instances") : null;
        if (instances != null)
        {
            for (Object instance : instances)
            {
                JsonObject instanceConfig = (JsonObject) instance;
                instanceConfigs.put((String) instanceConfig.get("id"), instanceConfig);
                LOGGER.debug("Loaded com.aws.proseve.ggcommons.config for {}", instanceConfig.get("id"));
            }
        }
    }

    abstract protected JsonObject loadConfiguration();

    abstract protected String getConfigSource();

    public JsonObject getGlobalConfig()
    {
        return globalConfig;
    }

    public Collection<String> getInstanceIds()
    {
        return instanceConfigs.keySet();
    }

    public JsonObject getInstanceConfig(String instanceId)
    {
        return instanceConfigs.getOrDefault(instanceId, null);
    }

    public TagConfiguration getTagConfig()
    {
        return tagConfig;
    }

    public HeartbeatConfiguration getHeartbeatConfig()
    {
        return heartbeatConfig;
    }

    public LoggingConfiguration getLoggingConfig()
    {
        return loggingConfig;
    }

    public String getThingName()
    {
        return thingName;
    }

    public String getComponentName()
    {
        return componentName;
    }

    public void addConfigChangeListener(ConfigurationChangeListener listener)
    {
        configChangeListeners.add(listener);
    }

    public void removeConfigChangeListener(ConfigurationChangeListener listener)
    {
        configChangeListeners.remove(listener);
    }

    protected void configurationChanged(JsonObject newConfig)
    {
        LOGGER.info("configurationChanged: Applying new com.aws.proseve.ggcommons.config: {}", newConfig);
        applyConfig(newConfig);

        LOGGER.info("configurationChanged: Notifying {} listeners", configChangeListeners.size());
        for (ConfigurationChangeListener listener : configChangeListeners)
        {
            if (listener != null) {
                listener.onConfigurationChanged();
            } else {
                LOGGER.error("ConfigurationChangeListener is null.  Not notifying.");
            }
        }
    }

    public String resolveTemplate(String template) {
        String retVal = template;
        if (template.contains("{ThingName}"))
        {
            retVal = retVal.replace("{ThingName}", getThingName());
        }
        if (template.contains("{ComponentName}"))
        {
            retVal = retVal.replace("{ComponentName}", getComponentName());
        }

        for (String tagKey : tagConfig.getKeys())
        {
            String hierarchyLevelTemplate = "{" + tagKey + "}";
            if (retVal.contains(hierarchyLevelTemplate))
            {
                retVal = retVal.replace(hierarchyLevelTemplate, tagConfig.getKeyValue(tagKey));
            }
        }
        return retVal;
    }

    protected JsonObject getDefaultConfig()
    {
        JsonObject retVal = new JsonObject();
        JsonObject logging = new JsonObject();
        JsonObject heartbeat = new JsonObject();
        JsonObject source = new JsonObject();
        JsonObject component = new JsonObject();
        JsonObject global = new JsonObject();
        JsonArray instances = new JsonArray();

        component.put("global", global);
        component.put("instances", instances);
        retVal.put("logging", logging);
        retVal.put("source", source);
        retVal.put("heartbeat", heartbeat);
        retVal.put("component", component);
        return retVal;
    }
}


